L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -1,110 +1,167 @@
-      HGCalTriggerDetId TC_idV9(detid);
+    for (auto& s: STCs){
 
-      int TC_wafer = TC_idV9.waferU() << kWafer_offset_ | TC_idV9.waferV();
-      int thickness = triggerTools_.thicknessIndex(detid);
+      int thickness = 0;
+      if ( triggerTools_.isSilicon(s.second.getSTCId() ) ){
+	thickness = triggerTools_.thicknessIndex(s.second.getSTCId(),true);
+      }else if ( triggerTools_.isScintillator(s.second.getSTCId()) ){
+	thickness = 3;
+      }
 
-      int TC_12th = 0;
-      int Uprime = 0;
-      int Vprime = 0;
-      int rocnum = detIdToROC_.getROCNumber(TC_idV9.triggerCellU(), TC_idV9.triggerCellV(), 1);
+      std::vector<uint32_t> output_ids = superTCmapping_.getConstituentTriggerCells( s.second.getSTCId() );
+
+      for (const auto& id: output_ids){
+
+	if ( fixedDataSizePerHGCROC_ && thickness > kHighDensityThickness_ && id!=superTCmapping_.getEvenDetId(id)) {
+	  continue;
+	}
+
+        if ( !triggerTools_.validTriggerCell(id) ) {
+          continue;
+        }
+        
+        l1t::HGCalTriggerCell triggerCell;
+
+        GlobalPoint point;
+	if ( fixedDataSizePerHGCROC_ == true && thickness > kHighDensityThickness_ ){
+	  point = coarseTCmapping_.getCoarseTriggerCellPosition( id );
+	}
+	else{
+	  point = triggerTools_.getTCPosition(id);
+	}
+        math::PtEtaPhiMLorentzVector p4(0, point.eta(), point.phi(), 0.);
+        triggerCell.setPosition(point);
+        triggerCell.setP4(p4);
+        triggerCell.setDetId(id);
+	trigCellVecOutput.push_back ( triggerCell );
+
+	if ( energyDivisionType_ == oneBitFraction ){ //Get the 1 bit fractions
+
+	  if ( id != s.second.getMaxId() ){
+	    float tc_fraction = getTriggerCellOneBitFraction( s.second.getTCpt(id) , s.second.getSumPt() );
+	    s.second.addToFractionSum(tc_fraction);
+	  }
+	}
 
-      if (rocnum == 1) {
-        Uprime = TC_idV9.triggerCellU();
-        Vprime = TC_idV9.triggerCellV() - TC_idV9.triggerCellU();
+      }
+    }
 
-      } else if (rocnum == 2) {
-        Uprime = TC_idV9.triggerCellU() - TC_idV9.triggerCellV() - 1;
-        Vprime = TC_idV9.triggerCellV();
+    // assign energy
+    for ( l1t::HGCalTriggerCell & tc : trigCellVecOutput){
+      const auto & stc = STCs[superTCmapping_.getCoarseTriggerCellId(tc.detId())]; 
+      assignSuperTriggerCellEnergy(tc, stc);        
+    }      
 
-      } else if (rocnum == 3) {
-        Uprime = TC_idV9.triggerCellU() - kRotate4_;
-        Vprime = TC_idV9.triggerCellV() - kRotate4_;
-      }
+}
 
-      TC_12th = (rocnum << kRocShift_) | ((Uprime << kUShift_ | Vprime) & kSplit_v9_);
+void
+HGCalConcentratorSuperTriggerCellImpl::
+ assignSuperTriggerCellEnergy(l1t::HGCalTriggerCell &c, const SuperTriggerCell &stc) const {
+      
+  if ( energyDivisionType_ == superTriggerCell ){
+    if ( c.detId() == stc.getMaxId() ){
+      c.setHwPt(stc.getSumHwPt());
+      c.setMipPt(stc.getSumMipPt());
+      c.setPt(stc.getSumPt());
+    }
+    else if ( c.detId() != stc.getMaxId() ){
+      c.setHwPt(0);
```

Comment:
What does a TC with 0 energy means? If it is a null TC it shouldn't appear in the output collection.

Reply:
It won't appear in the output collection, since there is a cut requiring energy > 0 in the ntuplizer. Since we now create all TCs, we need to have a way to remove the ones which are 0, I guess this cut is the best way.

Reply:
It is independent of the ntuplizer. The TPG simulation is run centrally, not the ntuplizer.
0 energy TCs should not appear in the collections produced by the producer.

Reply:
Ok - now the TC is not pushed-back if not the maximum (for the superTC case).

Reply:
Ok, so now this case where zeroes are filled should not happen. So an exception could be thrown in case for some unexpected reason the program enters here.